File: src/main/java/kr/co/bookand/backend/config/jwt/exception/ExceptionHandlerFilter.java
package kr.co.bookand.backend.config.jwt.exception;


import kr.co.bookand.backend.common.exception.ErrorCode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Slf4j
@Component
public class ExceptionHandlerFilter extends OncePerRequestFilter {
    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        try {
            filterChain.doFilter(request, response);
        } catch (JwtException ex) {
            setErrorResponse(HttpStatus.UNAUTHORIZED, response, ex.getErrorCode());
        } catch (RuntimeException ex) {
            log.error("runtime exception exception handler filter");
            setErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, response, ErrorCode.INTER_SERVER_ERROR);
        }
    }

    public void setErrorResponse(HttpStatus status, HttpServletResponse response, ErrorCode ex) {
        response.setStatus(status.value());
        response.setContentType("application/json");
        response.setCharacterEncoding("utf-8");
        JwtException errorResponse = new JwtException(ex, ex.getMessage());
        try {
            String json = errorResponse.convertToJson(ex.getMessage());
            response.getWriter().write(json);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/kr/co/bookand/backend/config/jwt/exception/ExceptionHandlerFilter.kt
package kr.co.bookand.backend.config.jwt.exception

import kr.co.bookand.backend.common.KotlinErrorCode
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class ExceptionHandlerFilter : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        try {
            filterChain.doFilter(request, response)
        } catch (ex: JwtException) {
            setErrorResponse(HttpStatus.UNAUTHORIZED, response, ex.errorCode)
        } catch (ex: RuntimeException) {
            setErrorResponse(HttpStatus.BAD_REQUEST, response, KotlinErrorCode.INTER_SERVER_ERROR)
        }
    }

    private fun setErrorResponse(
        status: HttpStatus,
        response: HttpServletResponse,
        ex: KotlinErrorCode
    ) {
        response.status = status.value()
        response.contentType = "application/json"
        response.characterEncoding = "utf-8"
        val errorResponse = JwtException(ex, ex.errorLog)
        try {
            val json = errorResponse.convertToJson(ex.errorLog)
            response.writer.write(json)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }


}